dissect/cstruct/expression.py

```diff
@@ -1,84 +1,298 @@
+
+    def hexdigit(self, token: str) -> bool:
+        return token in string.hexdigits
+
+    def operator(self, token: str) -> bool:
+        return token in {"*", "/", "+", "-", "%", "&", "^", "|", "(", ")", "~"}
+
+    def match(
+        self,
+        func: Optional[Callable[[str], bool]] = None,
+        expected: Optional[str] = None,
+        consume: bool = True,
+        append: bool = True,
+    ) -> bool:
+        if self.eol():
+            return False
+
+        token = self.get_token()
+
+        if expected and self.equal(token, expected):
+            if append:
+                self.tokens.append(token)
+            if consume:
+                self.consume()
+            return True
+
+        if func and func(token):
+            if append:
+                self.tokens.append(token)
+            if consume:
+                self.consume()
+            return True
+
+        return False
+
+    def consume(self) -> None:
+        self.pos += 1
+
+    def eol(self) -> bool:
+        return self.pos >= len(self.expression)
+
+    def get_token(self) -> str:
+        if self.eol():
+            raise ExpressionTokenizerError(f"Out of bounds index: {self.pos}, length: {len(self.expression)}")
+        return self.expression[self.pos]
+
+    def tokenize(self) -> list[str]:
+        token = ""
+
+        # Loop over expression runs in linear time
+        while not self.eol():
+            # If token is a single character operand add it to tokens
+            if self.match(self.operator):
+                continue
+
+            # If token is a single digit, keep looping over expression and build the number
+            elif self.match(self.digit, consume=False, append=False):
+                token += self.get_token()
+                self.consume()
+
+                # Support for binary and hexadecimal notation
+                if self.match(expected={"x", "X", "b", "B"}, consume=False, append=False):
+                    token += self.get_token()
+                    self.consume()
+
+                while self.match(self.hexdigit, consume=False, append=False):
+                    token += self.get_token()
+                    self.consume()
+                    if self.eol():
+                        break
+
+                # Checks for suffixes in numbers
+                if self.match(expected={"u", "U"}, consume=False, append=False):
+                    self.consume()
+                    self.match(expected={"l", "L"}, append=False)
+                    self.match(expected={"l", "L"}, append=False)
+
+                elif self.match(expected={"l", "L"}, append=False):
+                    self.match(expected={"l", "L"}, append=False)
+                    self.match(expected={"u", "U"}, append=False)
+                else:
+                    pass
+
+                # Number cannot end on x or b in the case of binary or hexadecimal notation
+                if len(token) == 2 and token[-1] in ("x", "X", "b", "B"):
```

Comment:
Maybe I did not understand the response to my previous question about this, but isn't it an idea to use a constant here?

i.e.

HEXBIN_SUFFIX = ("x", "X", "b", "B")

?

Reply:
Oh I forgot about that, apologies.